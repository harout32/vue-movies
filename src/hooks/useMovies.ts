import { ref, onMounted } from 'vue';
import MoviesApi from '../services/api/moviesApi';
import { Movie, MoviesQuery } from '../types/Movies';

export default function (perPage: number) {
  const isLoading = ref(true);
  const movies = ref<Movie[]>([]);
  const hasNext = ref<boolean>(false);

  const fetchMovies = async (query: MoviesQuery) => {
    isLoading.value = true;
    console.log('[NETWORK:GET_MOVIES:REQUEST]:', query);
    const { data } = await MoviesApi.getMovies({ ...query, perPage });
    console.log('[NETWORK:GET_MOVIES:RESPONSE]:', data);

    movies.value = data.list;
    hasNext.value = data.hasNext;
    isLoading.value = false;
  };

  onMounted(async () => {
    await fetchMovies({ pageIndex: 0, perPage });
  });

  return {
    movies,
    isLoading,
    hasNext,
    fetchMovies,
  };
}
