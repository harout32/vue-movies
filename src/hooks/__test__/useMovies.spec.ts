import useMovies from '../useMovies';

import MoviesApi from '../../services/api/moviesApi';

jest.mock('vue', () => {
  return {
    ...jest.requireActual('vue'),
    onMounted: jest.fn((cb) => cb()),
  };
});

describe('useMovies', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should call getMovies initially', () => {
    const spy = jest
      .spyOn(MoviesApi, 'getMovies')
      .mockImplementation(() => new Promise((resolve) => resolve({ data: { hasNext: false, list: [] } } as any)));

    useMovies(10);

    expect(spy).toHaveBeenCalledWith({ pageIndex: 0, perPage: 10 });
  });

  it('should call getMovies when calling fetchMovies', () => {
    const spy = jest
      .spyOn(MoviesApi, 'getMovies')
      .mockImplementation(() => new Promise((resolve) => resolve({ data: { hasNext: false, list: [] } } as any)));

    const { fetchMovies } = useMovies(10);

    fetchMovies({ filters: { genre: [1], showType: [0] }, pageIndex: 1, search: 'some search' });

    expect(spy.mock.calls).toMatchSnapshot();
  });
});
