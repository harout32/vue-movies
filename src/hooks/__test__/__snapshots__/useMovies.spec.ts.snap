// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMovies should call getMovies when calling fetchMovies 1`] = `
Array [
  Array [
    Object {
      "pageIndex": 0,
      "perPage": 10,
    },
  ],
  Array [
    Object {
      "filters": Object {
        "genre": Array [
          1,
        ],
        "showType": Array [
          0,
        ],
      },
      "pageIndex": 1,
      "perPage": 10,
      "search": "some search",
    },
  ],
]
`;
